sympy/matrices/expressions/diagonal.py

```diff
@@ -1,14 +1,19 @@
 from __future__ import print_function, division
 
 from sympy.matrices.expressions import MatrixExpr
-from sympy.core import S
-
+from sympy.core import S, Eq
+from sympy.functions.special.tensor_functions import KroneckerDelta
 class DiagonalMatrix(MatrixExpr):
     arg = property(lambda self: self.args[0])
     shape = property(lambda self: (self.arg.shape[0], self.arg.shape[0]))
 
     def _entry(self, i, j):
-        return S.Zero if i != j else self.arg[i, 0]
+        eq = Eq(i, j)
+        if eq is S.false:
+            return S.Zero
+        elif eq is S.true:
+            return self.arg[i, 0]
```

Comment:
I hope the tests fail, because it seems to me this should be `self.arg[i, i]`.

Also, I was thinking of adding a little beauty to the result by making the final return be

```
self.args[i, i]*KroneckerDelta(i,j)`. What do you think?